tests/test_surface_events.cpp

```diff
@@ -588,4 +588,52 @@ TEST_F(ClientSurfaceEventsTest, buffer_release)
     EXPECT_TRUE(buffer_released[2]);
 }
 
+TEST_F(ClientSurfaceEventsTest, surface_enters_output)
+{
+    using namespace testing;
+
+    wlcs::Client client{the_server()};
+
+    auto surface = client.create_visible_surface(100, 100);
+    client.roundtrip();
+
+    EXPECT_THAT(surface.current_outputs(), Not(IsEmpty())) << "Surface did not initially enter output";
+}
+
+TEST_F(ClientSurfaceEventsTest, surface_leaves_output)
+{
+    using namespace testing;
+
+    wlcs::Client client{the_server()};
+
+    auto surface = client.create_visible_surface(100, 100);
+    client.roundtrip();
+
+    ASSERT_THAT(surface.current_outputs(), Not(IsEmpty())) << "Precondition failed";
+
+    the_server().move_surface_to(surface, -200, 0);
```

Comment:
This assumes that the compositor will allow surfaces to be moved entirely off the visible area; I would hope that Mir, at least, would disallow this move.

Because WLCS is testing window-management code I think we want to allow compositors to use as much of their normal window management code as possible.

Reply:
This is unfortunate, but I suppose you are right